doc: use blockquotes for Stability: markers

addaleax opened the pull request:
Use blockquotes instead of code blocks for stability markers in
the docs. Doing that:

- Makes the makers appear correctly when viewed e.g. on github.
- Allows remark-lint rules like `no-undefined-references` to work
  properly (https://github.com/nodejs/node/pull/7729).

addaleax commented:
CI: https://ci.nodejs.org/job/node-test-commit/4130/

/cc @ChALkeR @nodejs/documentation

ChALkeR commented on doc/api/documentation.md:
```diff
@@ -37,34 +37,26 @@ and in the process of being redesigned.
+
+> Stability: 3 - Locked
+> Only fixes related to security, performance, or bug fixes will be accepted.
+> Please do not suggest API changes in this area; they will be refused.
```
This block doesn't look good on GitHub:
 1. This is interpreted as one blockquote, not four.
 2. Line breaks don't work.

Adding a manual line break there to fix how it looks on GitHub confuses the doctool and it treats those as separated blocks, with only the first line being colored.

ChALkeR commented:
Also, this change added extra margin to the stability markers in the html produced by the doc tool by wrapping these `<pre>` blocks in a blockquote. Can we somehow get rid of the blockquote or change its style?

As for the `documentation.md` changes — perhaps it would make sense to support both ways (code block and blockquote) and use the code block only in the `documentation.md` file?

ChALkeR commented:
Nice, thanks!

Just noticed: not sure why, but I have an `"undefined"` Text node under each stability level blockquote =). That doesn't affect the Stability code blocks in `documentation.md`.

addaleax commented:
@ChALkeR Thanks for noticing! Should be fixed now.

jasnell commented:
LGTM

ChALkeR commented:
@addaleax Something looks strange in the json output to me.

```js
{
          "textRaw": "fs.existsSync(path)",
          "type": "method",
          "name": "existsSync",
          "meta": {
            "added": [
              "v0.1.21"
            ],
            "deprecated": [
              "v1.0.0"
            ]
          },
          "desc": "<p>Stability: 0 - Deprecated: Use [<code>fs.statSync()</code>][] or [<code>fs.accessSync()</code>][]\ninstead.</p>\n<p>Synchronous version of [<code>fs.exists()</code>][].\nReturns <code>true</code> if the file exists, <code>false</code> otherwise.</p>\n",
          "signatures": [
            {
              "params": [
                {
                  "textRaw": "`path` {String | Buffer} ",
                  "name": "path",
                  "type": "String | Buffer"
                }
              ]
            },
            {
              "params": [
                {
                  "name": "path"
                }
              ]
            }
          ]
        },
```

Why are there two signatures there now?

addaleax commented:
Yeah, I was planning on double-checking the JSON output anyway, sorry.

ChALkeR commented on doc/api/domain.md:
```diff
@@ -436,7 +436,7 @@ without exiting the domain.
 ### domain.dispose()
 
-    Stability: 0 - Deprecated.  Please recover from failed IO actions
+> Stability: 0 - Deprecated.  Please recover from failed IO actions
```
The next line also should be converted to `>` or at least the padding should be removed. Else, this is how it renders:
![screenshot_20160720_200119](https://cloud.githubusercontent.com/assets/291301/16995405/e57efdf8-4eb4-11e6-85e0-b39237f6e637.png)

ChALkeR commented on doc/api/repl.md:
```diff
@@ -444,7 +444,7 @@ added: v2.0.0
 -->
 
-   Stability: 0 - Deprecated: Use `NODE_REPL_HISTORY` instead.
+> Stability: 0 - Deprecated: Use `NODE_REPL_HISTORY` instead.
```
Niiice. This fixed an actual error: https://nodejs.org/api/repl.html#repl_node_repl_history_file — that line had three spaces instead of four =).

ChALkeR commented:
Note: this also changes JSON output from being wrapped into `<pre><code>` into being wrapped into `<p>`.

Also, it generates now things like `[<code>fs.accessSync()</code>][]` instead of ``[`fs.accessSync()`][]`` (in the JSON). Btw, I'm not sure if ``[`fs.accessSync()`][]`` is correct there, shouldn't it be a link, as everything looks html-ish there? I'm not familiar with the json output.

ChALkeR commented:
Markdown — LGTM with one comment.
HTML — LGTM.
JSON — some comments above, but I'm not sure about most of those myself, so it shouldn't block this.

addaleax commented:
Rebased, and fixed the docs nit.

I’ll have to look into the JSON output myself, that one’s confusing… you’re right, there’s a difference here, but it looks like this change actually resulted in prepending the correct signature information where it was previously left out.

ChALkeR commented:
HTML output and *.md changes — LGTM
JSON — looks reasonable, but again — I'm not familiar with json output.
Perhaps /cc @Fishrock123?

Actual output changes here — https://gist.github.com/ChALkeR/9fc933caa0faca63ba5105b9bc453ec2.
HTML are fine that's just due to https://github.com/nodejs/node/pull/7757#discussion_r71566119 bugfix.

ChALkeR commented:
/cc @nodejs/collaborators

targos commented:
LGTM

silverwind commented:
Got some screenshots of the docs before/after? (sorry for being so lazy)

ChALkeR commented:
@silverwind There are no actual changes in the generated html apart of a bugfix at https://nodejs.org/api/repl.html#repl_node_repl_history_file, see the diff linked at https://github.com/nodejs/node/pull/7757#issuecomment-234902250.

GitHub render changes could be viewed by comparing https://github.com/nodejs/node/blob/master/doc/api/tls.md#deprecated-apis with https://github.com/addaleax/node/blob/doc-blockquote-stability/doc/api/tls.md#deprecated-apis, for example.

Screenshotting JSON output is pretty pointles.

addaleax commented:
Yeah, screenshots don’t really make sense here, and apart from that @ChALkeR’s diff is probably the best one can get for the JSON output. That one does _look_ okay to me.

silverwind commented:
Thanks. JSON looks alright, looks more correct than before at least, I'd say. LGTM.

addaleax commented:
Rebased to resolve conflict with 29e49fc286080215031a81effbd59eac092fff2f, new CI: https://ci.nodejs.org/job/node-test-commit/4393/

I’d like to land this if it’s green.

jasnell commented:
This still LGTM. Is this ready to go?

addaleax commented:
Yes, I’ll land it now (thanks for the ping!)

addaleax commented:
Landed in c809b8834541bf63b1a43d5a12833d98ffba834f, thanks for the reviews everyone!

TheAlphaNerd commented:
@addaleax this is not landing cleanly, likely due to a bunch of other commits needing to be backported. please feel free to manually backport
